### PR TITLE
fix(perception): Re-enable the perception of unequipped vehicles

### PR DESCRIPTION
**Pull request by schwepmo**
## Description

* With the integration of vehicle dimensions (i.e., length, width, height) #278 in the `VehicleObject`s the `CentralPerceptionComponent` required `VehicleRegistration`s to extract the dimensions
* This broke the ability for all vehicles to be handled in the perception index -> vehicles failing the `weight` condition in the `MappingAmbassador`
* With this PR this functionality was revived
* Additionally, all `VehicleType`s configured in the *mapping_config.json* are now held in the `vehicleTypes`-map in the `SimulationKernel`, which enables type retrieval when adding vehicles to the perception index
* vehicles with no configured prototype will have (0, 0, 0) as their dimensions
* the perception integration test has been extended to better look at edge cases

* future work may improve on this by adding actual `VehicleTypes` to the `ScenarioVehicleRegistration`-interaction, which would remove the requirement to define all external vehicle types as an additional `prototype` in the *mapping_config.json*

## Issue(s) related to this PR

* Resolves internal issue 551
  
## Affected parts of the online documentation

* We could discuss adding a section on the perception of *Scenario Vehicles*

## Definition of Done

**Prerequisites**

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] Your GitHub user id is linked with your Eclipse Account.

**Required**

- [x] The title of this merge request follows the scheme `type(scope): description`  (in the style of [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary))
- [x] You have assigned a suitable label to this pull request (e.g., `enhancement`, or `bugfix`)
- [x] `origin/main` has been merged into your Fork.
- [x] Coding guidelines have been followed (see CONTRIBUTING.md).
- [x] All checks on GitHub pass.
- [x] All tests on [Jenkins](https://ci.eclipse.org/mosaic/job/mosaic/) pass.

**Requested** (can be enforced by maintainers)

- [ ] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [x] If a bug has been fixed, a new unit test has been written (beforehand) to prove misbehavior
- [x] There are no new SpotBugs warnings.

### CODE REVIEW

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ApplicationAmbassador.java [hunk 1]
```diff
@@ -678,7 +682,6 @@ private void addVehicleIfNotYetAdded(long time, String unitName) {
         final VehicleRegistration vehicleRegistration = vehicleRegistrations.remove(unitName);
         if (vehicleRegistration != null) {
             UnitSimulator.UnitSimulator.registerVehicle(time, vehicleRegistration);
-            SimulationKernel.SimulationKernel.getCentralPerceptionComponent().addVehicle(vehicleRegistration);
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ApplicationAmbassador.java:681]: Why can we remove this?

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/providers/VehicleTree.java [hunk 2]
```diff
@@ -76,13 +76,24 @@ public List<VehicleObject> getVehiclesInRange(PerceptionModel searchRange) {
...
 
+    @Override
+    protected VehicleObject addOrGetVehicle(VehicleData vehicleData) {
+        boolean isNew = !indexedVehicles.containsKey(vehicleData.getName());
+        VehicleObject vehicleObject = super.addOrGetVehicle(vehicleData);
+        if (isNew) {
+            vehicleTree.addItem(vehicleObject);
+        }
+        return vehicleObject;
+    }
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/providers/VehicleTree.java:87]: This code exists three times now, maybe a better solution could be:

* define a new abstract method `void onVehicleAdded(VehicleObject v)` in class `VehicleIndex` and call this method whenever a vehicle object added to the index.
* implement the method in `VehicleGrid`, `VehicleTree`, ... and add the VehicleObject to the spatial index.

File: fed/mosaic-mapping/src/main/java/org/eclipse/mosaic/fed/mapping/ambassador/MappingAmbassador.java [hunk 3]
```diff
@@ -131,28 +133,28 @@ private void handleInteraction(ScenarioVehicleRegistration scenarioVehicle) thro
...
                         prototype.weight
                 );
-                return;
+                applications = new ArrayList<>();
             }
 
             final VehicleRegistration vehicleRegistration = new VehicleRegistration(
                     scenarioVehicle.getTime(),
                     scenarioVehicle.getName(),
                     prototype.group,
```
Comment (kschrab) [inline fed/mosaic-mapping/src/main/java/org/eclipse/mosaic/fed/mapping/ambassador/MappingAmbassador.java:150]: I would think that the group should also only be used from the prototype on the same conditions as the list of applications. In the case the weight condition is false, we should pass `null` instead. (?)

**Review by kschrab — CHANGES_REQUESTED** [review]

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ApplicationAmbassador.java (hunk 1 above)
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ApplicationAmbassador.java:681]: This method was only required to fetch the vehicle type from the vehicle registration before it is updated and moved.
Now the type information is cached as soon as the vehicle registration is received in `process(final VehicleRegistration vehicleRegistration)`, which removes the necessity of the `addVehicle`-method

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/providers/VehicleTree.java (hunk 2 above)
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/providers/VehicleTree.java:87]: Good point! Actually same applies for update and removal operations allowing to only implement `updateVehicles(...)` in the `VehicleIndex` and add the abstract methods `onIndexUpdate` and `onVehicleRemoval`

Comment (kschrab) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/providers/VehicleTree.java]: image

**Review by kschrab — APPROVED** [review]
